src/libraries/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -199,34 +199,43 @@ private static unsafe void ProcessResult(SocketError errorCode, GetAddrInfoExCon
         {
             GetAddrInfoExState state = GetAddrInfoExState.FromHandleAndFree(context->QueryStateHandle);
 
+            object result;
             try
             {
                 CancellationToken cancellationToken = state.UnregisterAndGetCancellationToken();
 
                 if (errorCode == SocketError.Success)
                 {
                     IPAddress[] addresses = ParseAddressInfoEx(context->Result, state.JustAddresses, out string? hostName);
-                    state.SetResult(state.JustAddresses ? (object)
+                    result = state.JustAddresses ?
                         addresses :
                         new IPHostEntry
                         {
                             HostName = hostName ?? state.HostName,
                             Aliases = Array.Empty<string>(),
                             AddressList = addresses
-                        });
+                        };
                 }
                 else
                 {
                     Exception ex = (errorCode == (SocketError)Interop.Winsock.WSA_E_CANCELLED && cancellationToken.IsCancellationRequested)
-                        ? (Exception)new OperationCanceledException(cancellationToken)
+                        ? new OperationCanceledException(cancellationToken)
                         : new SocketException((int)errorCode);
-                    state.SetResult(ExceptionDispatchInfo.SetCurrentStackTrace(ex));
+                    result = ExceptionDispatchInfo.SetCurrentStackTrace(ex);
                 }
             }
+            catch (Exception ex)
+            {
+                result = ex;
+            }
             finally
             {
                 state.Dispose();
             }
+
+            // GetAddrInfoExState is a SafeHandle and Dispose() will only release its' GetAddrInfoExContext pointer;
+            // the rest of the object's state is still valid and the instance will be used as an IThreadPoolWorkItem.
+            state.SetResult(result);
```

Comment:
IMO it would be clearer if `state` had a method like

```c#
public void ReleaseContext() => Dispose();
```

and this comment was there. This would generally save the confusion of thinking "why is there Dispose before SetResult? is that even ok?" and then having to read the comment each time.